File: src/selfupdate.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"time"
)

const updateRepo = "sethcarney/skl"
const releasesAPI = "https://api.github.com/repos/" + updateRepo + "/releases/latest"

type githubAsset struct {
	Name               string `json:"name"`
	BrowserDownloadURL string `json:"browser_download_url"`
}

type githubRelease struct {
	TagName string        `json:"tag_name"`
	Assets  []githubAsset `json:"assets"`
}

func getBinaryAssetName() string {
	switch runtime.GOOS {
	case "linux":
		switch runtime.GOARCH {
		case "amd64":
			return "skl-linux-x64"
		case "arm64":
			return "skl-linux-arm64"
		}
	case "darwin":
		switch runtime.GOARCH {
		case "amd64":
			return "skl-macos-x64"
		case "arm64":
			return "skl-macos-arm64"
		}
	case "windows":
		if runtime.GOARCH == "amd64" {
			return "skl-windows-x64.exe"
		}
	}
	return ""
}

func isNewer(latest, current string) bool {
	lParts := strings.Split(strings.TrimPrefix(latest, "v"), ".")
	cParts := strings.Split(strings.TrimPrefix(current, "v"), ".")
	for i := 0; i < 3; i++ {
		var l, c int
		if i < len(lParts) {
			l, _ = strconv.Atoi(lParts[i])
		}
		if i < len(cParts) {
			c, _ = strconv.Atoi(cParts[i])
		}
		if l > c {
			return true
		}
		if l < c {
			return false
		}
	}
	return false
}

func runSelfUpdate(currentVersion string) {
	client := &http.Client{Timeout: 30 * time.Second}

	fmt.Printf("%sChecking for updates...%s\n", ansiDim, ansiReset)

	req, _ := http.NewRequest("GET", releasesAPI, nil)
	req.Header.Set("User-Agent", "skl-cli/"+currentVersion)
	resp, err := client.Do(req)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to check for updates: %v\n", err)
		os.Exit(1)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		fmt.Fprintf(os.Stderr, "GitHub API returned %d\n", resp.StatusCode)
		os.Exit(1)
	}

	body, _ := io.ReadAll(resp.Body)
	var release githubRelease
	if err := json.Unmarshal(body, &release); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to parse release info: %v\n", err)
		os.Exit(1)
	}

	latestVersion := strings.TrimPrefix(release.TagName, "v")

	if !isNewer(latestVersion, currentVersion) {
		fmt.Printf("%sAlready up to date%s %s(%s)%s\n", ansiText, ansiReset, ansiDim, currentVersion, ansiReset)
		return
	}

	fmt.Printf("%sNew version available:%s %s %s(current: %s)%s\n",
		ansiText, ansiReset, latestVersion, ansiDim, currentVersion, ansiReset)
	fmt.Println()

	assetName := getBinaryAssetName()
	if assetName == "" {
		fmt.Fprintf(os.Stderr, "Unsupported platform: %s/%s\n", runtime.GOOS, runtime.GOARCH)
		fmt.Fprintf(os.Stderr, "Download manually from: https://github.com/%s/releases/latest\n", updateRepo)
		os.Exit(1)
	}

	var downloadURL string
	for _, a := range release.Assets {
		if a.Name == assetName {
			downloadURL = a.BrowserDownloadURL
			break
		}
	}
	if downloadURL == "" {
		fmt.Fprintf(os.Stderr, "Binary for your platform (%s) not found in release %s.\n", assetName, latestVersion)
		os.Exit(1)
	}

	fmt.Printf("%sDownloading %s...%s\n", ansiDim, assetName, ansiReset)
	dlResp, err := client.Get(downloadURL)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Download failed: %v\n", err)
		os.Exit(1)
	}
	defer dlResp.Body.Close()

	if dlResp.StatusCode != 200 {
		fmt.Fprintf(os.Stderr, "Download failed: HTTP %d\n", dlResp.StatusCode)
		os.Exit(1)
	}

	dlBody, _ := io.ReadAll(dlResp.Body)
	tmpPath := filepath.Join(os.TempDir(), fmt.Sprintf("skl-update-%d", time.Now().UnixNano()))

	if err := os.WriteFile(tmpPath, dlBody, 0755); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write temp file: %v\n", err)
		os.Exit(1)
	}

	execPath, err := os.Executable()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Could not determine executable path: %v\n", err)
		os.Exit(1)
	}

	if runtime.GOOS != "windows" {
		if err := os.Rename(tmpPath, execPath); err != nil {
			// Cross-filesystem: unlink + copy
			os.Remove(execPath)
			if err2 := copyFile(tmpPath, execPath); err2 != nil {
				fmt.Fprintf(os.Stderr, "Failed to update binary: %v\n", err2)
				os.Exit(1)
			}
			os.Remove(tmpPath)
		}
		fmt.Printf("%sUpdated to %s successfully.%s\n", ansiText, latestVersion, ansiReset)
		fmt.Printf("%sRestart your shell or run %sskl --version%s%s to confirm.%s\n",
			ansiDim, ansiText, ansiDim, ansiReset, ansiReset)
	} else {
		batchPath := filepath.Join(os.TempDir(), "skl-update.bat")
		batchContent := fmt.Sprintf("@echo off\r\ntimeout /t 1 /nobreak > NUL\r\nmove /y \"%s\" \"%s\" > NUL\r\ndel \"%%~f0\"\r\n",
			tmpPath, execPath)
		if err := os.WriteFile(batchPath, []byte(batchContent), 0644); err != nil {
			fmt.Fprintf(os.Stderr, "Failed to write update script: %v\n", err)
			os.Exit(1)
		}
		if err := exec.Command("cmd", "/c", "start", "/b", "", batchPath).Start(); err != nil {
			fmt.Fprintf(os.Stderr, "Failed to launch update script: %v\n", err)
			fmt.Printf("%sTo apply the update manually, run:%s\n  %s%s%s\n",
				ansiDim, ansiReset, ansiText, batchPath, ansiReset)
			return
		}
		fmt.Printf("%sUpdated to %s successfully.%s\n", ansiText, latestVersion, ansiReset)
		fmt.Printf("%sApplying update as this process exits...%s\n", ansiDim, ansiReset)
		os.Exit(0)
	}
}

File: legacy /src/self-update.ts
import { chmodSync, copyFileSync, renameSync, unlinkSync, writeFileSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';

const REPO = 'sethcarney/skills';
const RELEASES_API = `https://api.github.com/repos/${REPO}/releases/latest`;

interface GitHubAsset {
  name: string;
  browser_download_url: string;
}

interface GitHubRelease {
  tag_name: string;
  assets: GitHubAsset[];
}

function getBinaryAssetName(): string | null {
  const { platform, arch } = process;
  if (platform === 'linux' && arch === 'x64') return 'skills-linux-x64';
  if (platform === 'linux' && arch === 'arm64') return 'skills-linux-arm64';
  if (platform === 'darwin' && arch === 'x64') return 'skills-macos-x64';
  if (platform === 'darwin' && arch === 'arm64') return 'skills-macos-arm64';
  if (platform === 'win32' && arch === 'x64') return 'skills-windows-x64.exe';
  return null;
}

// When compiled with Bun, the Bun global is available; in Node/npx it is not.
function isRunningAsBinary(): boolean {
  return typeof (globalThis as Record<string, unknown>)['Bun'] !== 'undefined';
}

function isNewer(latest: string, current: string): boolean {
  const parse = (v: string) => v.split('.').map(Number);
  const [l, c] = [parse(latest), parse(current)];
  for (let i = 0; i < 3; i++) {
    const diff = (l[i] ?? 0) - (c[i] ?? 0);
    if (diff > 0) return true;
    if (diff < 0) return false;
  }
  return false;
}

export async function runSelfUpdate(currentVersion: string): Promise<void> {
  const RESET = '\x1b[0m';
  const DIM = '\x1b[38;5;102m';
  const TEXT = '\x1b[38;5;145m';

  console.log(`${DIM}Checking for updates...${RESET}`);

  let release: GitHubRelease;
  try {
    const res = await fetch(RELEASES_API, {
      headers: { 'User-Agent': `skills-cli/${currentVersion}` },
    });
    if (!res.ok) throw new Error(`GitHub API returned ${res.status}`);
    release = (await res.json()) as GitHubRelease;
  } catch (err) {
    console.error(
      `Failed to check for updates: ${err instanceof Error ? err.message : String(err)}`
    );
    process.exit(1);
  }

  const latestVersion = release.tag_name.replace(/^v/, '');

  if (!isNewer(latestVersion, currentVersion)) {
    console.log(`${TEXT}Already up to date${RESET} ${DIM}(${currentVersion})${RESET}`);
    return;
  }

  console.log(
    `${TEXT}New version available:${RESET} ${latestVersion} ${DIM}(current: ${currentVersion})${RESET}`
  );
  console.log();

  if (!isRunningAsBinary()) {
    console.log(
      `${DIM}You are running skills via npx or npm. To get self-updates, install the binary:${RESET}`
    );
    console.log();
    console.log(
      `  ${TEXT}curl -fsSL https://raw.githubusercontent.com/${REPO}/main/install.sh | sh${RESET}`
    );
    console.log();
    return;
  }

  const assetName = getBinaryAssetName();
  if (!assetName) {
    console.error(`Unsupported platform: ${process.platform}/${process.arch}`);
    console.error(`Download manually from: https://github.com/${REPO}/releases/latest`);
    process.exit(1);
  }

  const asset = release.assets.find((a) => a.name === assetName);
  if (!asset) {
    console.error(`Binary for your platform (${assetName}) not found in release ${latestVersion}.`);
    console.error(`Check: https://github.com/${REPO}/releases/tag/v${latestVersion}`);
    process.exit(1);
  }

  console.log(`${DIM}Downloading ${assetName}...${RESET}`);
  const downloadRes = await fetch(asset.browser_download_url);
  if (!downloadRes.ok) {
    console.error(`Download failed: ${downloadRes.status}`);
    process.exit(1);
  }

  const buffer = Buffer.from(await downloadRes.arrayBuffer());
  const tmpPath = join(tmpdir(), `skills-update-${Date.now()}`);
  writeFileSync(tmpPath, buffer);

  const currentBinary = process.execPath;

  if (process.platform !== 'win32') {
    chmodSync(tmpPath, 0o755);
    try {
      renameSync(tmpPath, currentBinary);
    } catch (err: unknown) {
      if ((err as NodeJS.ErrnoException).code === 'EXDEV') {
        // Unlink the running binary first so copyFileSync can write to the path
        // without hitting ETXTBSY (text file busy). The running process retains
        // its inode and continues normally.
        unlinkSync(currentBinary);
        copyFileSync(tmpPath, currentBinary);
        unlinkSync(tmpPath);
      } else {
        throw err;
      }
    }
    console.log(`${TEXT}Updated to ${latestVersion} successfully.${RESET}`);
    console.log(
      `${DIM}Restart your shell or run ${TEXT}skills --version${DIM} to confirm.${RESET}`
    );
  } else {
    // Windows cannot replace a running executable directly.
    // Write a helper batch file the user can run after exiting.
    const batchPath = join(tmpdir(), 'skills-update.bat');
    writeFileSync(
      batchPath,
      `@echo off\r\ntimeout /t 1 /nobreak > NUL\r\nmove /y "${tmpPath}" "${currentBinary}" > NUL\r\ndel "%~f0"\r\n`
    );
    console.log(`${TEXT}Downloaded ${latestVersion} to:${RESET} ${tmpPath}`);
    console.log(`${DIM}To apply the update, run after exiting this process:${RESET}`);
    console.log(`  ${TEXT}${batchPath}${RESET}`);
  }
}
